Analytics

rooftopcellist opened the pull request:
##### SUMMARY
The Insights Analytics feature packages up information about an Ansible Tower Instance/Cluster


##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request



##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - Installer

##### TOWER VERSION
tower_version=3.5.0


##### ADDITIONAL INFORMATION
To test this out, you must:
* have the insights-client installed
* register a valid RH account with the subscription-manager

```
docker exec -u 0 -it tools_awx_1 bash
awx-manage gather_analytics
tar xfvz <file.tar.gz>

#  will produce the following:
./counts.json
./config.json
./projects_by_scm_type.json

```

ryanpetrello committed:
add a minimal framework for generating analytics/metrics

annotate queries & add license analytics

benthomasson commented:
Please add %z to the end of strftime in the file name of the tarball.

rooftopcellist committed:
add basic job & JT analytics

ryanpetrello commented on awx/api/views/root.py:
```diff
@@ -157,6 +158,7 @@ def get(self, request, format=None):
             'ha': is_ha_environment(),
             'version': get_awx_version(),
             'active_node': settings.CLUSTER_HOST_ID,
+            'system_uuid': settings.system_uuid,
```
Does `SYSTEM_UUID` need to be capitalized?

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,168 @@
+@register('job_counts')   # TODO: evaluate if we want this (was not an ask)  Also, may think about annotating rather than grabbing objects for efficiency 
+def job_counts(since):    # TODO: Optimize -- for example, all of these are going to need to be restrained to the last 24 hours/INSIGHTS_SCHEDULE
+    counts = {}
+    counts['total_jobs'] = models.UnifiedJob.objects.all().count()
```
Let's figure out how to express these _nearly_ identical queries in the ORM as _one_ query grouped by status i.e.,

```sql
awx=> SELECT COUNT(id), status FROM main_unifiedjob GROUP BY status;
 count |   status
-------+------------
   352 | failed
  2964 | successful
     2 | error
    10 | canceled
(4 rows)
```

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,168 @@
+        counts[instance.id] = {'uuid': instance.uuid,
+                               'jobs_total': instance.jobs_total,       # this is _all_ jobs run by that node
+                               'jobs_running': instance.jobs_running,   # this is jobs in running & waiting state
+                               'launch_type': {'manual': models.UnifiedJob.objects.filter(launch_type='manual').count(), # I can definitely condense this
```
These queries under `launch_type` don't seem to be actually *restricted* to the instance group in question - I assume we want these to be the jobs that _ran on_ these instances (i.e., `main_unifiedjob.execution_node = main_instance.hostname`)?

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,168 @@
+    counts['pending_jobs'] = models.UnifiedJob.objects.filter(status='pending').count()
+    counts['waiting_jobs'] = models.UnifiedJob.objects.filter(status='waiting').count()
+    counts['running_jobs'] = models.UnifiedJob.objects.filter(status='running').count()
+    for instance in models.Instance.objects.all():
```
This loop means a _lot_ of nested queries.  Let's find a way to use to the ORM to turn this into one or two aggregate queries, which could provide us with raw data we could work with.  Something like:

```sql
SELECT main_instance.uuid, COUNT(main_unifiedjob.id), main_unifiedjob.launch_type FROM main_unifiedjob, main_instance WHERE main_unifiedjob.execution_node=main_instance.hostname GROUP BY main_instance.id, main_unifiedjob.launch_type;
                 uuid                 | count | launch_type
--------------------------------------+-------+-------------
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |  1047 | sync
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |     1 | scm
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   807 | dependency
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |    31 | relaunch
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   185 | manual
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   228 | workflow
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   830 | scheduled
(7 rows)
```

```sql
SELECT main_instance.uuid, COUNT(main_unifiedjob.id), main_unifiedjob.status FROM main_unifiedjob, main_instance WHERE main_unifiedjob.execution_node=main_instance.hostname GROUP BY main_instance.id, main_unifiedjob.status;
                 uuid                 | count |   status
--------------------------------------+-------+------------
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |     6 | canceled
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |     1 | error
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |   326 | failed
 ec2c225e-c00d-5ef8-71c6-0bdd2490d6d2 |  2796 | successful
(4 rows)
```

ryanpetrello replied:
From this data, we could do some math to _calculate_ `Instance.jobs_running` (it's the total of the count column, grouped by uuid), and `Instance.jobs_running` (it's the value of the row with `status='running'` or `0` if there is no row).

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,168 @@
+    ])
+
+    counts['active_host_count'] = models.Host.objects.active_count()
+    counts['smart_inventories'] = models.Inventory.objects.filter(kind='smart').count()
```
These two queries could be *one* aggregate query where we count and group by `kind`.

ryanpetrello commented on awx/api/views/__init__.py:
```diff
@@ -317,6 +317,54 @@ def get(self, request, format=None):
+        data.append("# TYPE sessions.expired_sessions counter")
+        data.append("sessions.expired_sessions {0}".format(expired_sessions))
+        
+        return Response(_prepare_data(data))
```
You could probably replace the `prepare_data` function with just:

```python
"\n".join(metrics)
```

rooftopcellist replied:
nice, that's a good point.  I'll be taking this whole section out eventually, this is just test data from when I was trying to hook up Prometheus to test things out.  I'll be writing a separate function to prepare the data from the analytics data for the /metrics endpoint, but I'll split that off into another PR

benthomasson commented:
Please test and report performance in seconds with 10, 100, 1000, 10,000, 100,000, 1,000,000, and etc entries in the appropriate tables like job events for queries added in this PR.   Plotting the results is a bonus requirement.

rooftopcellist committed:
add insights setting, optimize and consolidate queries

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -126,47 +124,24 @@ def projects_by_scm_type(since):
-@register('job_counts_instance')
-def job_counts_instance(since):
-    counts = {}
+    counts['status'] = dict(models.UnifiedJob.objects.values_list('status').annotate(Count('status')))
```
👍

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -126,47 +124,24 @@ def projects_by_scm_type(since):
-                               }
+                                'jobs_total': models.UnifiedJob.objects.filter(execution_node=instance.hostname, status__in=('running', 'waiting',)).count(),
+                                'jobs_running': models.UnifiedJob.objects.filter(execution_node=instance.hostname).count(),   # jobs in running & waiting state
+                                'launch_type': dict(models.UnifiedJob.objects.filter(execution_node=instance.hostname).values_list('launch_type').annotate(Count('launch_type')))
```
👍

ryanpetrello replied:
If you've got e.g., 25 instances, this is 25 aggregate queries.  Is there a way to do this all as *one* query grouped by `(main_instance.hostname, main_unifiedjob.launch_type)` similar to 
https://github.com/ansible/awx/pull/3284/commits/85002052c9a97e10148e2289b1bc60e54ddcc9b6#r261036977?

rooftopcellist commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,147 @@
+def inventory_counts(since):
+    counts = {}
+    from django.db.models import Count
+    for inv in models.Inventory.objects.annotate(num_sources=Count('inventory_sources', distinct=True), 
```
I found a way to simplify this:

```
list(Inventory.objects.filter(kind='smart').values('id').annotate(num_sources=Count('inventory_sources'), 
                                            num_hosts=Count('hosts')).values('id', 'kind', 'name', 'num_sources', 'num_hosts'))[0]
```

Unfortunately, I also found that smart inventories use a different manager meaning all of the host counts for smart_inventories are zero unless I grab each smart_inventory object, so I'll have to do the smart_inventories in a separate query.

rooftopcellist committed:
add copy events table and remove events queries

add copy unifiedjob table, futher optimization, rm since from job_instance queries

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -60,16 +60,18 @@ def counts(since):
-    counts['inventories'] = dict(models.Inventory.objects.order_by().values_list('kind').annotate(Count('kind'))) # TODO: s/''/'normal'
-
+    inv_counts = dict(models.Inventory.objects.order_by().values_list('kind').annotate(Count('kind')))
+    inv_counts['normal'] = inv_counts[''] # rename '' to 'normal'
```
maybe `inv_counts.pop('')` so there's not a duplicate?

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -0,0 +1,173 @@
+    return decorate
+
+
+def copy_events_table(path='/tmp'):
```
Organization-wise, let's put this function into `collectors.py`  It's not *technically* using the `@register` decorator, but it's the same thing in spirit - it's taking database content and generating data/a file.

rooftopcellist replied:
good point, I'll move that to the collectors to keep all data collection methods grouped together.

jbradberry commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,178 @@
+        counts[org['id']] = {'name': org['name'],
+                            'users': org['num_users'],
+                            'teams': org['num_teams']
+                            }
```
I'm pretty sure that this type of indentation is going to fail when checked by flake8.

jbradberry commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,178 @@
+    for smart_inv in models.Inventory.objects.filter(kind='smart'):
+        counts[smart_inv.id] = {'name': smart_inv.name,
+                          'kind': smart_inv.kind,
+                          'num_hosts': smart_inv.hosts.count(),
```
Don't forget to give these the same treatment as the annotation above.  In fact, it looks like it could be done in a single loop.

rooftopcellist replied:
thanks for catching that. I'll add that to my PR.

jbradberry replied:
Oh, I think I see what you mean about smart inventories and hosts that you mentioned in one of the other comments.  I'd push for making a change to that through-table so that you can do the counts as an annotation.

ryanpetrello replied:
@chrismeyersfsu can you weigh in on this performance-wise?

rooftopcellist replied:
@wwitzel3 took a look at this with me and did query analysis on this as is compared to it as one consolidated query and it performed better as is.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    ])
+
+    inv_counts = dict(models.Inventory.objects.order_by().values_list('kind').annotate(Count('kind')))
+    inv_counts['normal'] = inv_counts[''] 
```
You can consolidate these two lines i.e.,

`inv_counts['normal'] = inv_counts.pop('')`

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+
+    
+@register('org_counts')
+def org_counts(since):
```
This function needs a test, see 
https://github.com/ansible/awx/pull/3284/files#diff-ea83b94c4e87b2a5361aeadf86f8a005 😄 

For all of the functions in this module, let's write some tests that make _real_ objects and verify that we get back the correct totals we anticipate (that way, we *know* this code works).

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    
+    
+@register('cred_type_counts')
+def cred_type_counts(since):
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    
+    
+@register('inventory_counts')
+def inventory_counts(since):
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+
+
+@register('projects_by_scm_type')
+def projects_by_scm_type(since):
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    return counts
+
+@register('instance_info')   # maybe integrate this with job_instance_counts?
+def instance_info(since):
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    return info
+
+
+@register('job_counts')
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    
+    
+@register('job_instance_counts')
+def job_instance_counts(since):         #TODO: all of these are going to need to be restrained to the last 24 hours/INSIGHTS_SCHEDULE
```
Needs a test.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+def cred_type_counts(since):
+    counts = {}
+    for cred_type in models.CredentialType.objects.annotate(num_credentials=Count('credentials', distinct=True)):  
+        counts[cred_type.id] = {'name': cred_type.name,
```
Might be valuable to add `managed_by_tower` here so we can filter out customer-provided credential types in our analytics.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+    
+@register('job_instance_counts')
+def job_instance_counts(since):         #TODO: all of these are going to need to be restrained to the last 24 hours/INSIGHTS_SCHEDULE
+    counts = {}
```
Are we still planning to limit this function on the "last 24h" basis cc @matburt @wenottingham ?

wenottingham replied:
We should gather since last run. If there is no last run, we should probably gather 2 days or so.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+def copy_events_table(path='/tmp/'):
+    events_file = os.path.join(path, 'events_table.csv')
+    try:
+        write_data = open(events_file2, 'w')
```
`events_file2` `NameError`

rooftopcellist replied:
that was a troubleshooting artifact.  I am abstracting a `copy_table()` method out and using it for an event and job table drops.  and need to add since, rather than hard coding 24 hours.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.failed, 
+                                      main_jobevent.changed, 
+                                      main_jobevent.uuid, 
+                                      main_jobevent.playbook, 
```
@matburt @wenottingham do we need this column?  You could _probably_ argue that this **could** constitute personally identifiable information.

wenottingham replied:
The usage case is:
- find long running / often failing (outlier) tasks
- sum usage of modules/roles
- track deprecation warnings

We should collect the minimum set of data that allows for that.

wenottingham replied:
The issue I could see that would require some sort of playbook/play/role/task gathering would be how you'd refer back to the outlier once you find it.

ryanpetrello replied:
I understand, I just want to be clear that while limiting what we collect here _could_ hamper our ability to report, this data _could_ also represent PII.  So we have a judgement call to make.

wenottingham replied:
Just to be clear: if there are fields here that are not relevant to performing those calculations, etc - we should update this code to not collect them. I am not specifically reviewing the field list.

ryanpetrello replied:
@wenottingham @rooftopcellist I've thought about this, and I think we _shouldn't_ capture `main_jobevent.task` and `main_jobevent.play`, because they're very likely to contain free-form text and potentially PII.

I think `main_jobevent.playbook` is probably pretty important, though, so we can communicate with users *which* specific playbooks might be using Ansible poorly.

I also think `main_jobevent.role` is pretty useful.

So to summarize from this list, I'd say let's lose `play`, `task`, `modified`, `uuid`.

rooftopcellist replied:
Ok, I've removed those fields, as well as:
```
                              main_jobevent.counter, 
                              main_jobevent.start_line, 
                              main_jobevent.end_line, 
                              main_jobevent.parent_id, 
                              main_jobevent.parent_uuid ```

matburt replied:
I left a comment here:

https://github.com/ansible/awx/pull/3284/files#r267120036

I think if we exclude this data then we'll lose a lot of dimensionality. We *can* convert this information to something else on the backend... hash it or something, but we will lose the ability to drill into any sort of detail when analyzing events across playbook and job runs.

matburt replied:
As useful as playbook data is, more and more detail is pushed down to roles (and thus plays). Having play/task information lets us correlate roles and plays even *between* job templates and being able to detect when plays change position under a playbook and track performance and characteristics in a larger scope.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.playbook, 
+                                      main_jobevent.play, 
+                                      main_jobevent.role, 
+                                      main_jobevent.task, 
```
Same question about `task` and `play`.

rooftopcellist replied:
resolved in earlier comment ^^. Removing `task` and `play`

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.end_line, 
+                                      main_jobevent.job_id, 
+                                      main_jobevent.host_id, 
+                                      main_jobevent.host_name, 
```
Same question re: `host_name`.  Do we _need_ this?

rooftopcellist replied:
If we identify problems connecting to certain hosts, we wouldn't be able to return which hosts were exhibiting issues.  But there is a strong argument against including host_name's from a PII perspective.  @ryanpetrello @wenottingham

wenottingham replied:
1. I don't see how you give information on flaky hosts without the host
2. Anyone using insights is already sending all their RHEL host information to RH

rooftopcellist replied:
ok, that makes sense, leaving the host_name in.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.parent_id, 
+                                      main_jobevent.parent_uuid 
+                                      FROM main_jobevent 
+                                      WHERE main_jobevent.created > NOW()- INTERVAL '1 DAY'
```
@rooftopcellist this is _probably_ fine for now, but if we ever make the beat interval configurable, we need to change this.

ryanpetrello replied:
Actually, it would probably be better to pass in a `since` timestamp here and just use it, no?

rooftopcellist replied:
yep, I'll change that to pass `since`

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -0,0 +1,142 @@
+    :pararm module: the module to search for registered analytic collector
+                    functions; defaults to awx.main.analytics.collectors
+    """
+    import time                 # TODO: Remove this
```
Let's go ahead and remove this and any other `TODO` where we're printing.

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -0,0 +1,142 @@
+                    os.remove(f.name)
+
+    post_query_time = time.time()
+    print("Analytics Total Query Time --- %s seconds ---" % (post_query_time - start_time))  # TODO: Remove this
```
Let's lose it.

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -0,0 +1,142 @@
+        last_run = max_interval
+
+    if _valid_license() is False:
+        logger.exception("Invalid License provided, or No License Provided")
```
@one-t let's verify that these actually show up in logs in these scenarios (i.e., that we've got the log configuration correct).

one-t replied:
Confirmed, these are showing up in logs sent to logstash

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+def copy_events_table(path='/tmp/'):
+    events_file = os.path.join(path, 'events_table.csv')
+    try:
+        write_data = open(events_file2, 'w')
```
Do we need to declare an encoding (`utf-8`) when we open this?  @one-t @rooftopcellist let's make sure we test this functionality w/ wacky unicode content.

jbradberry replied:
Aren't we still supporting Python 2?  I don't believe even 2.7 supports the Python 3-style `open()` with encoding parameter.

ryanpetrello replied:
Nope, awx only supports/works in Python3 now.

jbradberry replied:
Really?  Excellent, go crazy then.

rooftopcellist replied:
It looks like in Python3, the `open()` method defaults to `utf-8` encoding.  https://docs.python.org/3/howto/unicode.html#the-string-type

ryanpetrello replied:
https://docs.python.org/3/library/functions.html#open

> encoding is the name of the encoding used to decode or encode the file. This should only be used in text mode. The default encoding is platform dependent (whatever locale.getpreferredencoding() returns), but any text encoding supported by Python can be used. See the codecs module for the list of supported encodings

```python
>>> type(open('/tmp/tmp5lppOW', 'r', encoding='utf-8').read())
<class 'str'>
>>> type(open('/tmp/tmp5lppOW', 'r').read())
<class 'str'>
>>> import locale
<module 'locale' from '/usr/lib64/python3.6/locale.py'>
>>> locale.getpreferredencoding()
'UTF-8'
```

👍

ryanpetrello commented on awx/main/migrations/0064_v350_analytics_state.py:
```diff
@@ -0,0 +1,25 @@
+        ('main', '0063_v350_org_host_limits'),
+    ]
+
+    operations = [
```
Be aware, you're probably going to be racing @AlanCoding to claim this one, I expect he'll beat you in https://github.com/ansible/awx/pull/3266

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.event, 
+                                      main_jobevent.failed, 
+                                      main_jobevent.changed, 
+                                      main_jobevent.uuid, 
```
These uuids are probably worthless _to us_ and probably don't need to be collected.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+        with connection.cursor() as cursor:
+            cursor.copy_expert('''COPY (SELECT main_jobevent.id, 
+                                      main_jobevent.created, 
+                                      main_jobevent.modified, 
```
`modified` is probably not useful/actionable in and of itself, because *anything* can technically modify the row in the table.  `created` definitely is.

rooftopcellist replied:
ok, removing modified

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+        write_data = open(events_file2, 'w')
+        with connection.cursor() as cursor:
+            cursor.copy_expert('''COPY (SELECT main_jobevent.id, 
+                                      main_jobevent.created, 
```
@rooftopcellist there's a number of *keys* under `main_jobevent.event_data` that we *probably* want to capture here, but we _do not_ want the entire data structure.  What if we added some columns like this one:

```
SELECT main_jobevent.id,
    ...
    event_data::json->'task_action',
```

That said, @matburt @chrismeyersfsu @benthomasson this probably throws a *notable* performance wrench in because this column is `TEXT`, not `jsonb`, so we're going to have to parse these (which is going to potentially be slow/expensive past 24h).  That said, if we *really want* this data, we don't really have any other options (we _can't_ send the entire field value, because it potentially contains a _lot_ of sensitive data, and it's going to be hellish to convert this to jsonb w/ a migration).

rooftopcellist replied:
@ryanpetrello What other keys from `main_jobevent.event_data` might we want?

Might we also want things from main_unifiedjob.job_args?

ryanpetrello replied:
No, historically speaking, `job_args` will _very_ likely contain sensitive data if people aren't properly using Tower's credential secret support (i.e., pasting plaintext passwords in places they don't belong).  Let's collect the _bare minimum_ necessary to deliver the things @wenottingham has asked for.

ryanpetrello replied:
Specifically:

> What modules am I utilizing, and how much?  if 90% of your tasks are shell/command, it should gently chide you

This is where `event_data::json->'task_action'` likely comes in.

...and:

> Are there anomalies in automation time for certain hosts, or certain playbooks?

@jladdjr I expect your perf plugin work will be most helpful here ^  Without that, we don't really have this type of data, historically.

> Are there anomalies in automation success for certain hosts or certain playbooks (flaky hosts/playbooks)

I think at a bare minimum here, we _probably_ want to capture *these* columns:

![image](https://user-images.githubusercontent.com/214912/54640712-d6a88a00-4a66-11e9-8032-f5067a3bbea3.png)


> How many people are using my corporate-approved roles, and how often?

`main_jobevent.role` should cover this for us, I think.

jladdjr replied:
> Are there anomalies in automation time for certain hosts, or certain playbooks?
>> @jladdjr I expect your perf plugin work will be most helpful here ^ Without that, we don't really have this type of data, historically.

@ryanpetrello - yes, definitely 👍 That sounds like it's right up the perf callback plugin's ally

jbradberry commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+
+    job_types = models.UnifiedJob.objects.values_list(
+        'execution_node', 'launch_type').annotate(job_launch_type=Count('launch_type'))
+    for job in job_types:
```
For readability's sake, you should probably unpack the variables in the tuple for these loops, something like

`for node, launch_type, count in job_types:`

rooftopcellist replied:
@jbradberry I nested `launch_type` and `status` under each execution node.  Is this what you meant?  For example:

```Out[31]: 
{'awx': 
     {
        'launch_type': {'successful': 6},
        'status': {'dependency': 1, 'manual': 4, 'sync': 1}
      }
}
```

jbradberry replied:
I just meant that in updating `counts` in the line below, you're referring to `job[0]`, `job[1]`, and `job[2]`.  It'd be nice for future readers of the code for those variables to have names.

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -0,0 +1,142 @@
+    print("Analytics Total Query Time --- %s seconds ---" % (post_query_time - start_time))  # TODO: Remove this
+
+    # copies Job Events data from db as a csv
+    collectors.copy_events_table(dest)
```
@rooftopcellist let's wrap this function call with some exception handling and logging so that if it throws an exception for some reason (what if the database goes down for a split second?  more likely: what if we introduce a subtle bug into the function), it doesn't break _everything_.  Similar to:

https://github.com/ansible/awx/pull/3284/files#diff-e58aed5ef9fc8ef40758f220398f0adaR98

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,218 @@
+                                      main_jobevent.play, 
+                                      main_jobevent.role, 
+                                      main_jobevent.task, 
+                                      main_jobevent.counter, 
```
Probably also don't _need_ line counting (`counter`, `start_line`, `end_line`).

rooftopcellist replied:
removed

rooftopcellist commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+                                 main_unifiedjob.polymorphic_ctype_id,
+                                 django_content_type.model,
+                                 main_unifiedjob.created, 
+                                 main_unifiedjob.description, 
```
we can probably lose the following fields:
* description
* celery_task_id
* created_by_id
* old_pk

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+        return events_file
+
+    events_query = '''COPY (SELECT main_jobevent.id, 
+                              main_jobevent.created, 
```
Let's add `event_data::json->'task_args'` to this list.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+                              main_jobevent.role, 
+                              main_jobevent.job_id, 
+                              main_jobevent.host_id, 
+                              main_jobevent.host_name, 
```
Let's add `event_data::json->'task_action'` to this list so we can track module usage.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+                                 main_unifiedjob.job_explanation, 
+                                 main_unifiedjob.start_args, 
+                                 main_unifiedjob.result_traceback, 
+                                 main_unifiedjob.celery_task_id, 
```
I think we should lose:

`start_args` (has had sensitive data leaks, historically; also, in more modern AWX versions it's encrypted anyways)
`celery_task_id` (useless info outside of the context of a running AWX system)
`result_traceback` (can be very verbose free-form text, it's not actionable or easily parseable, it _could_ contain sensitive context)

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+                                 main_unifiedjob.unified_job_template_id, 
+                                 main_unifiedjob.launch_type, 
+                                 main_unifiedjob.schedule_id, 
+                                 main_unifiedjob.execution_node, 
```
`execution_node` will be neat/interesting 😄

rooftopcellist replied:
We will already have the counts from `job_instance_counts.json` earlier (done with queries).  Now that we have this, some of those queries are somewhat redundant.

rooftopcellist replied:
I left in the job_instance_counts(), but took out the `last_run/since` stipulation.  If this is performant enough, we can use it for metrics.

matburt commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,236 @@
+            return e
+        return events_file
+
+    events_query = '''COPY (SELECT main_jobevent.id, 
```
Can we add play, task, uuid ,and parent_uuid?

Otherwise we'll never be able to do summarization under task or play nor will we be able to correlate task or play changes between different runs (cross-run analysis).

I would think, in order to do effective automation analysis, that we need to be able to identify task and play information in some way across automation.

Otherwise the best we'll be able to analyze is overall job runtime information, and possibly per-host information... but we'd never be able to drill down into analytics that might identify problematic tasks/roles/plays/playbooks.

ryanpetrello replied:
Originally I asked to remove `play` and `task` given:

> Goal is to not collect PII (personally identifiable information), or secrets.

But if we've decided we _can_ and _must_ collect this data, then I guess it's fine.

rooftopcellist replied:
ok, play, task, uuid, and parent_uuid have been added back.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,237 @@
+                              main_jobevent.uuid,
+                              main_jobevent.parent_uuid,
+                              main_jobevent.event, 
+                              main_jobevent.event_data::json->'task_action'
```
@matburt do you have any performance concerns about this?  This column is `TEXT`, so we have to convert it to a json type to do this.  I think this is probably fine, though, considering we're scoping this copy to the last 24 hours.

matburt replied:
This is fine, I don't have any concerns here.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+        'license_expiry': license_info.get('time_remaining', 0),
+        'pendo_tracking': settings.PENDO_TRACKING_STATE,
+        'authentication_backends': settings.AUTHENTICATION_BACKENDS,
+        'logging_aggregators': settings.LOG_AGGREGATOR_LOGGERS
```
This is interesting data:`settings.LOG_AGGREGATOR_LOGGERS`:

```
In [2]: settings.LOG_AGGREGATOR_LOGGERS
Out[2]: ['awx', 'activity_stream', 'job_events', 'system_tracking']
```

But we probably also want to know if logging is **enabled**, and if so, what type they're using:

```python
'external_logger_enabled': settings.LOG_AGGREGATOR_ENABLED,
'external_logger_type': getattr(settings, 'LOG_AGGREGATOR_TYPE', None),
```

AlanCoding replied:
yeah, list is totally irrelevant if it's not enabled. That's only for sub-selection in case it is enabled.

rooftopcellist replied:
Ok, I figure Insights will check if it is enabled first, and if not, then they can just disregard the `LOG_AGGREGATOR_LOGGERS` setting.  

I've added this, here is an example of how it will show with loggly set up:

```
Out[103]: 
{'ansible_version': '2.7.8',
 'authentication_backends': ['awx.sso.backends.TACACSPlusBackend',
  'django.contrib.auth.backends.ModelBackend'],
 'external_logger_enabled': True,
 'external_logger_type': 'loggly',
 'free_instances': 0,
 'license_expiry': 8519115645,
 'license_type': 'enterprise',
 'logging_aggregators': ['awx', 'activity_stream', 'job_events'],
 'pendo_tracking': 'off',
 'system_uuid': '00000000-0000-0000-0000-000000000000',
 'tower_url_base': 'https://localhost:8043',
 'tower_version': '3.0.1-282-gd2dcd1186'}
```

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+        counts[camelcase_to_underscore(cls.__name__)] = cls.objects.count()
+
+    venvs = get_custom_venv_choices()
+    counts['custom_virtualenvs'] = len([
```
@one-t @rooftopcellist let's work w/ @vismay-golwala and make sure that when his custom venv path feature merges, those venvs *also* show up in this list (looking at his PR, I expect they would)

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+                'capacity': instance['capacity'],
+                'cpu': instance['cpu'],
+                'memory': instance['memory'],
+                'managed_by_policy': instance['managed_by_policy'],
```
`last_isolated_check` might be useful for tracking isolated instance usage

ryanpetrello replied:
`enabled` might also be interesting to have

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+@register('job_counts')
+def job_counts(since):
+    counts = {}
+    counts['total_jobs'] = models.UnifiedJob.objects.all().count()
```
Let's *definitely* lose this `.all()` or we'll have serious performance issues for larger users.

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+def job_instance_counts(since):
+    counts = {}
+    from django.utils.timezone import now, timedelta
+    since = now() - timedelta(days=1)
```
Isn't a `since` value already being passed in?  Why are we calculating it again?  Are we _using_ this value in this function?

rooftopcellist replied:
yes, I've "since" removed this. pushing soon

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+                cursor.copy_expert(query, write_data)
+                write_data.close()
+        except Exception as e:
+            return e
```
We're returning `e` here, but we're not doing anything with the return value at https://github.com/ansible/awx/pull/3284/files#diff-742f36c0cca9a5137d76c5a198af227cR213 and https://github.com/ansible/awx/pull/3284/files#diff-742f36c0cca9a5137d76c5a198af227cR236.  If an exception occurs, it's going to get swallowed up and not be visible.  Let's let any exceptions here be raised upwards (by removing this `try...except` block), and let the caller of `copy_tables()` catch them (which you're already doing here: https://github.com/ansible/awx/pull/3284/files#diff-e58aed5ef9fc8ef40758f220398f0adaR93)

ryanpetrello commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,238 @@
+                              ORDER BY main_jobevent.id ASC) to stdout'''.format(since.strftime("'%Y-%m-%d %H:%M:%S'"))
+    _copy_table(table='events', query=events_query, path=full_path)
+    
+    unified_job_query = '''COPY (SELECT DISTINCT main_unifiedjob.id, 
```
Is there a need for `DISTINCT`?  You're joining on:

```sql
main_unifiedjob.polymorphic_ctype_id = django_content_type.id
```

I don't anticipate a scenario where a `unified_job` row would have _multiple_ references to different content type IDs (job, inventory update, etc...).
In other words, this relationship *should* be one-to-one, so you should only be getting back N rows where N is the total count of `unified_job` (created since yesterday).

rooftopcellist replied:
ok, that makes sense.  that will make it a little more performant too.

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -290,6 +290,16 @@ def _load_default_license_from_file():
+    field_class=fields.BooleanField,
+    default=False,
+    label=_('Insights Analytics Data'),
+    help_text=_('Enables Tower to gather analytics data about itself and send it to Insights.'),
```
@wenottingham do we want to explicitly spell this out as "Red Hat Insights" or something else?

wenottingham replied:
```
label=_('Gather data for Automation Insights')
help_text=_('Enables Tower to gather data on automation and send it to Red Hat Insights.')
```

ryanpetrello commented on awx/main/conf.py:
```diff
@@ -290,6 +290,16 @@ def _load_default_license_from_file():
     placeholder={'HTTP_PROXY': 'myproxy.local:8080'},
 )
 
+register(
```
Where is this flag actually *enforced*?  Did you mean to update this?

https://github.com/ansible/awx/pull/3284/files#diff-9d4ea1dd908b35fb92eaede4bd10bb46R326

rooftopcellist replied:
No, we enforce it on the `INSIGHTS_DATA_ENABLED` setting, not sure how that got lost.

rooftopcellist committed:
update setting wording & add instance-enabled field

rooftopcellist committed:
update setting wording & add instance-enabled field

Commit:
add ui for insights enablement

ryanpetrello commented:
recheck

rooftopcellist commented:
It looks like implicit project_updates are being included in the unifiedjob counts.  After talking with @one-t and @matburt, we concluded that we probably want to exclude `implicit project_updates`.  I need to figure out how to distinguishing implicit from explicit and how to exclude implicit project updates.

ryanpetrello commented:
@rooftopcellist I think `type='run'` is the implicit ones:

https://github.com/ansible/awx/blob/devel/awx/main/tasks.py#L1529

```
awx-dev=> SELECT job_type FROM main_projectupdate;
 job_type
----------
 check
 run
 check
 run
 check
 run
(6 rows)
```

@AlanCoding @matburt have I got this right, or backwards?

AlanCoding commented:
Yes, this is correct. The "run" wording that the git module actually syncs the file tree on the system. The "check" version is a "check" job by Ansible terminology - meaning that it find the revision of the project, but does not update it.

It's kind of backwards of how it works big-picture in Tower, because getting the revision means that a permanent change is made to the Project model, updating which revision will be used in playbook runs, and "run" has an ephemeral feeling, that it only assures the local source tree matches the expected revision.

I don't know about query performance implications or anything like that.

matburt commented:
Just to confirm...`run` is the implicit and `check` are the explicit runs.

matburt commented:
The `check` vs `run` is more technically correct because it's talking about actions taken by `Ansible` itself.

rooftopcellist commented:
ok, thanks for the info ^^.  

I just synced up with @bender and we will also need:
* Add Table Copy of Job Templates, make sure to include Modules (in jobevents) -- and make sure to be able to count modules.  

Also, this would be useful for /metrics:
* Add a query for number of runs for Job Templates

rooftopcellist committed:
check for open license

ryanpetrello commented on awx/main/analytics/core.py:
```diff
@@ -25,6 +26,8 @@
 
 def _valid_license():
     try:
+        if get_license(show_key=False)['license_type'] == 'open':
```
If there's no license, `license_type` won't be a valid key.  Use:

`.get('license_type', 'UNLICENSED')`

rooftopcellist committed:
collect jt table for analytics

rooftopcellist committed:
exclude implicit project updates

chrismeyersfsu commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,258 @@
+    
+    counts['active_host_count'] = models.Host.objects.active_count()   
+    active_sessions = Session.objects.filter(expire_date__gte=now()).count()
+    api_sessions = models.UserSessionMembership.objects.select_related('session').filter(session__expire_date__gte=now()).count()
```
make sure expire_date is indexed.

rooftopcellist replied:
`expire_date` is indexed:

```Indexes:
    "django_session_pkey" PRIMARY KEY, btree (session_key)
    "django_session_expire_date_a5c62663" btree (expire_date)
    "django_session_session_key_c0390e0f_like" btree (session_key varchar_pattern_ops)
```

chrismeyersfsu commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,258 @@
+@register('org_counts')
+def org_counts(since):
+    counts = {}
+    for org in models.Organization.objects.annotate(num_users=Count('member_role__members', distinct=True), 
```
make a bunch of users that belong to Organization (20,000 users)

chrismeyersfsu commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,258 @@
+def job_counts(since):
+    counts = {}
+    counts['total_jobs'] = models.UnifiedJob.objects.count()
+    counts['status'] = dict(models.UnifiedJob.objects.values_list('status').annotate(Count('status')))
```
Note that `status` doesn't have an index.

rooftopcellist replied:
I think it this case the performance gained by a `status` and `launch_type` for the read for analytics data is not worth the performance lost for create, update, delete for unified_jobs.

rooftopcellist replied:
On second thought, we don't generally do any updates or deletes on unified jobs and there aren't too many creates relatively (each job run there are ~2).  I will investigate adding an index for `status` and `launch_type` post-merge in a follow-up issue.

chrismeyersfsu commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,258 @@
+    counts = {}
+    counts['total_jobs'] = models.UnifiedJob.objects.count()
+    counts['status'] = dict(models.UnifiedJob.objects.values_list('status').annotate(Count('status')))
+    counts['launch_type'] = dict(models.UnifiedJob.objects.values_list('launch_type').annotate(Count('launch_type')))
```
No index on `launch_type`

chrismeyersfsu commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,258 @@
+    for job in job_types:
+        counts.setdefault(job[0], {}).setdefault('status', {})[job[1]] = job[2]
+        
+    job_statuses = models.UnifiedJob.objects.values_list(
```
Let me tinker on this query. I _think_ the number of results returned is going to be # of jobs. I think it can't be the number of unique pairs of <execution_node, status> instead.

chrismeyersfsu replied:
This is fine :+1:

Ladas commented on awx/main/analytics/collectors.py:
```diff
@@ -0,0 +1,260 @@
+        'tower_version': get_awx_version(),
+        'ansible_version': get_ansible_version(),
+        'license_type': license_info.get('license_type', 'UNLICENSED'),
+        'free_instances': license_info.get('free instances', 0),
```
@rooftopcellist hello, would it be possible to add here the total_instances count? So we can compute percentage remaining...

rooftopcellist replied:
@Ladas I believe the key for that is "instance_count".  That shouldn't be too difficult.
